storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3574,14 +3599,138 @@ dberr_t fsp_sys_tablespace_validate()
+uint32_t fseg_get_max_extent(const fseg_header_t *header, mtr_t *mtr)
+{
+  buf_block_t *iblock;
+  fseg_inode_t *inode= fseg_inode_try_get(header, 0, 0, mtr, &iblock);
+  const uint16_t ioffset= uint16_t(inode - iblock->page.frame);
+  uint32_t max_used_extent= 0;
+  fseg_get_n_frag_pages(inode, &max_used_extent);
+  uint32_t used_extent= flst_get_max_extent(
+    iblock, uint16_t(ioffset + FSEG_NOT_FULL), mtr);
+
+  max_used_extent= std::max(max_used_extent, used_extent);
+  used_extent= flst_get_max_extent(
+    iblock, uint16_t(ioffset + FSEG_FULL), mtr);
+
+  max_used_extent= std::max(max_used_extent, used_extent);
+  used_extent= flst_get_max_extent(
+    iblock, uint16_t(ioffset + FSEG_FREE), mtr);
+
+  max_used_extent= std::max(max_used_extent, used_extent);
+  return max_used_extent;
+}
+
+/** Get the maximum used extent in the system tablespace
+@param sys_root_pages  all system tablespace index root page
+@param max_used_extent maximum used extent
+@return error code */
+dberr_t fsp_get_max_sys_used_index_extent(
+  const std::vector<uint32_t> &sys_root_pages,
+  uint32_t &max_used_extent)
+{
+  dberr_t err= DB_SUCCESS;
+  mtr_t mtr;
+  for (auto page : sys_root_pages)
+  {
+    mtr.start();
+    buf_block_t *block= buf_page_get_gen(page_id_t{0, page}, 0, RW_S_LATCH,
+      nullptr, BUF_GET, &mtr, &err);
+    if (!block)
+    {
+      mtr.commit();
+      return err;
+    }
+
+    max_used_extent= std::max(max_used_extent,
+                              get_extent_start_page(page));
+    /* Get non-leaf segment pages */
+    fseg_header_t *seg_header= block->page.frame + PAGE_HEADER
+                               + PAGE_BTR_SEG_TOP;
+    ut_a(mach_read_from_4(seg_header + FSEG_HDR_SPACE) == 0);
+    uint32_t seg_page_no= mach_read_from_4(
+      seg_header + FSEG_HDR_PAGE_NO);
+    max_used_extent= std::max(max_used_extent,
+                              get_extent_start_page(seg_page_no));
+
+    max_used_extent= std::max(max_used_extent,
+                              fseg_get_max_extent(seg_header, &mtr));
+
+    /* Get leaf segment pages */
+    seg_header= block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_LEAF;
+    ut_a(mach_read_from_4(seg_header + FSEG_HDR_SPACE) == 0);
+    seg_page_no= mach_read_from_4(seg_header + FSEG_HDR_PAGE_NO);
+
+    max_used_extent= std::max(max_used_extent,
+                              get_extent_start_page(seg_page_no));
+
+    max_used_extent= std::max(max_used_extent,
+                              fseg_get_max_extent(seg_header, &mtr));
+    mtr.commit();
+  }
+
+  return err;
+}
+
 void fsp_system_tablespace_truncate()
 {
+  dberr_t err= DB_SUCCESS;
+  uint32_t max_used_extent= 0;
+  if (srv_sys_shrink_extensive)
+  {
+    std::vector<uint32_t> sys_root_pages;
+    err= dict_get_sys_tablespace_indexes(sys_root_pages);
+    if (err) return;
+
+    max_used_extent= std::max(max_used_extent, buf_dblwr.block2_page());
+    fsp_get_max_sys_used_index_extent(sys_root_pages, max_used_extent);
```

Comment:
I thought that extents and page numbers are not the same. It would be good to add some comments here.

What about the indexes of the hard-coded system tables `SYS_TABLES`, `SYS_COLUMNS`, `SYS_INDEXES` and `SYS_FIELDS`? Should they be included in `sys_root_pages`?